server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregationBuilder.java

```diff
@@ -167,38 +223,46 @@ static Rounding createRounding(Rounding.DateTimeUnit interval, ZoneId timeZone)
     @Override
     protected XContentBuilder doXContentBody(XContentBuilder builder, Params params) throws IOException {
         builder.field(NUM_BUCKETS_FIELD.getPreferredName(), numBuckets);
+        builder.field(MINIMUM_INTERVAL_FIELD.getPreferredName(), minimumIntervalExpression);
         return builder;
     }
 
     @Override
     protected int innerHashCode() {
-        return Objects.hash(numBuckets);
+        return Objects.hash(numBuckets, minimumIntervalExpression);
     }
 
     @Override
     protected boolean innerEquals(Object obj) {
         AutoDateHistogramAggregationBuilder other = (AutoDateHistogramAggregationBuilder) obj;
-        return Objects.equals(numBuckets, other.numBuckets);
+        return Objects.equals(numBuckets, other.numBuckets) && Objects.equals(minimumIntervalExpression, other.minimumIntervalExpression);
     }
 
     public static class RoundingInfo implements Writeable {
         final Rounding rounding;
         final int[] innerIntervals;
         final long roughEstimateDurationMillis;
         final String unitAbbreviation;
+        final String dateTimeUnit;
 
-        public RoundingInfo(Rounding rounding, long roughEstimateDurationMillis, String unitAbbreviation, int... innerIntervals) {
+        public RoundingInfo(Rounding rounding,
+                            long roughEstimateDurationMillis,
+                            String unitAbbreviation,
+                            String dateTimeUnit,
+                            int... innerIntervals) {
             this.rounding = rounding;
             this.roughEstimateDurationMillis = roughEstimateDurationMillis;
             this.unitAbbreviation = unitAbbreviation;
             this.innerIntervals = innerIntervals;
+            this.dateTimeUnit = dateTimeUnit;
```

Comment:
can we `assert dateTimeUnit != null;` here so we definitely catch the case where the `allowedIntervals` map doesn't contain an entry for the date time unit used in the rounding?

Reply:
addressed in 95ee1cf0320a901e70efa41a3c9f9785d5aeb59f